### PR TITLE
Improve pre-aggregation for olap scan node with join node

### PR DESCRIPTION
**Pull request by Youngwb**
#245 
1. Cross join cannot pre-aggregation
2. For other types of join, only one side can be pre-aggregation

### CODE REVIEW

**Comment by HangyuanLiu** [comment]
Can you add a comment on how to confirm whether pre-aggregation can be turned on, and add it at the beginning of PreAggregateTurnOnRule.java to facilitate other developers to understand the logic of preaggregation in the plan

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java [hunk 1]
```diff
@@ -272,15 +276,50 @@ private boolean checkAggregations(PreAggregationContext context, PhysicalOlapSca
...
+                context.notPreAggregationJoin = true;
+                context.groupings.clear();
+                context.aggregations.clear();
+                return visit(optExpression, context);
+            }
+
+            // For other types of joins, only one side can turn on pre aggregation
+            ColumnRefSet usedColumns = new ColumnRefSet();
+            context.groupings.forEach(g -> usedColumns.union(g.getUsedColumns()));
+            context.aggregations.forEach(a -> usedColumns.union(a.getUsedColumns()));
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java:299]: I think aggregation columns don't need only from one of children

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java [hunk 2]
```diff
@@ -272,15 +276,50 @@ private boolean checkAggregations(PreAggregationContext context, PhysicalOlapSca
...
+            boolean checkLeft = leftOutputColumns.contains(usedColumns);
+            boolean checkRight = rightOutputColumns.contains(usedColumns);
+
+            PreAggregationContext disableContext = new PreAggregationContext();
+            disableContext.notPreAggregationJoin = true;
+
+            if (checkLeft) {
+                leftChild.getOp().accept(this, leftChild, context);
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java:307]: context.groupings is need add Join on clause columns?

Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java:334]: joinPredicates add Join on clause columns now

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java (hunk 1 above)
Comment (Youngwb) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/implementation/PreAggregateTurnOnRule.java]: Fixed

**Comment by Youngwb** [comment]
> Can you add a comment on how to confirm whether pre-aggregation can be turned on, and add it at the beginning of PreAggregateTurnOnRule.java to facilitate other developers to understand the logic of preaggregation in the plan

Done

**Review by Seaven — DISMISSED** [review]

**Review by kangkaisen — APPROVED** [review]

**Review by Seaven — APPROVED** [review]

**Pull request merged**